src/NServiceBus.Metrics/RawData/RingBuffer.cs

```diff
@@ -0,0 +1,106 @@
+﻿namespace NServiceBus.Metrics.RawData
+{
+    using System;
+    using System.Threading;
+
+    class RingBuffer
+    {
+        public const int Size = 4096;
+        const long SizeMask = Size - 1;
+        const long EpochMask = ~SizeMask;
+
+        long nextToWrite;
+        long nextToConsume;
+
+        Entry[] entries = new Entry[Size];
+
+        public struct Entry
+        {
+            public Entry(long ticks, long value)
+            {
+                Ticks = ticks;
+                Value = value;
+            }
+
+            public long Ticks;
+            public long Value;
+        }
+
+        public bool TryWrite(long value)
+        {
+            var readNextWrite = Volatile.Read(ref nextToWrite);
+            long index;
+
+            do
+            {
+                var consume = Volatile.Read(ref nextToConsume);
+
+                // if writers overlaps with not consumed writes, end
+                if (readNextWrite - consume >= Size)
```

Comment:
A valid point @mikeminutillo . Let me try to address it.

The method provide `TryWrite` semantics and is up to the caller whether they want to `SpinWait` for a free slot or give up the write. This has not been wired up with `Probes` yet. With a good reporter (and I believe that the one included in this PR is good) and a big enough buffer (currently `4096`), the overlap is highly unlikely to occur. The reporter algorithm is pretty simple, scheduling next consume right away if anything was read from the buffer, see: https://github.com/Particular/NServiceBus.Metrics/pull/33/files/6474df9a76ca61f800d803205ccc27aadfa53bd3#r124200923